table/src/main/java/com/bazaarvoice/emodb/table/db/astyanax/AstyanaxTableDAO.java

```diff
@@ -346,12 +384,37 @@ public MaintenanceOp apply(Storage storage) {
      * Helper returns the next maintenance operation that should be performed on the specified table uuid/storage.
      */
     @Nullable
-    private MaintenanceOp getNextMaintenanceOp(final TableJson json, final Storage storage) {
+    private MaintenanceOp getNextMaintenanceOp(final TableJson json, final Storage storage, Supplier<List<TableEventDatacenter>> tableEventDatacenterSupplier) {
         final StorageState from = storage.getState();
         switch (from) {
-            case PRIMARY:
-                return null;  // Nothing to do.  This is the common case.
+            case PRIMARY: {
+                List<TableEvent> tableEvents = getTableEvents(json.getTable(), storage.getUuidString(), tableEventDatacenterSupplier.get(), TableEvent.Action.PROMOTE);
+
+                // Nothing to do.  This is the common case.
+                if (tableEvents.isEmpty()) {
+                    return null;
+                }
+
+                Instant maxTimestamp = tableEvents.stream()
+                        .map(TableEvent::getEventTime)
+                        .map(TimeUUIDs::getTimeMillis)
+                        .max(Long::compare)
+                        .map(Instant::ofEpochMilli)
+                        .get();
 
+                Instant when = maxTimestamp.plus(MIN_CONSISTENCY_DELAY);
+
+                return MaintenanceOp.forMetadata("TableEvent:mark-ready", when, new MaintenanceTask() {
+                    @Override
+                    public void run(Runnable progress) {
+
+                        checkPlacementConsistent(_systemTablePlacement, maxTimestamp);
+                        checkPlacementConsistent(storage.getPlacement(), maxTimestamp);
+
+                        checkForAnyNonCompleteTableEvents(json.getTable(), storage.getUuidString(), TableEvent.Action.PROMOTE);
```

Comment:
This is where we are marking the events as ready for promote side (for TABLE TEMPLATE and PLACEMENT MOVE actions). correct me if I'm wrong, as I'm not sure when we hit the PRIMARY maintenance.

Reply:
exactly, we do promote after the new data is available in the primary state